src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -4262,126 +4241,567 @@ bool CallData::CheckResolutionLocked(grpc_call_element* elem,
+      subchannel_call_->StartTransportStreamOpBatch(batch);
+    }
+    return;
+  }
+  // Add the batch to the pending list.
+  PendingBatchesAdd(batch);
+  // Check if we've already gotten a subchannel call.
+  // Note that once we have picked a subchannel, we do not need to acquire
+  // the channel's data plane mutex, which is more efficient (especially for
+  // streaming calls).
+  if (subchannel_call_ != nullptr) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_client_channel_call_trace)) {
+      gpr_log(GPR_INFO,
+              "chand=%p lb_call=%p: starting batch on subchannel_call=%p",
+              chand_, this, subchannel_call_.get());
+    }
+    PendingBatchesResume();
+    return;
+  }
+  // We do not yet have a subchannel call.
+  // For batches containing a send_initial_metadata op, acquire the
+  // channel's data plane mutex to pick a subchannel.
+  if (GPR_LIKELY(batch->send_initial_metadata)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_client_channel_call_trace)) {
+      gpr_log(GPR_INFO,
+              "chand=%p lb_call=%p: grabbing data plane mutex to perform pick",
+              chand_, this);
+    }
+    PickSubchannel(this, GRPC_ERROR_NONE);
+  } else {
+    // For all other batches, release the call combiner.
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_client_channel_call_trace)) {
+      gpr_log(GPR_INFO,
+              "chand=%p lb_call=%p: saved batch, yielding call combiner",
+              chand_, this);
+    }
+    GRPC_CALL_COMBINER_STOP(call_combiner_,
+                            "batch does not include send_initial_metadata");
+  }
+}
+
+void LoadBalancedCall::RecvTrailingMetadataReadyForLoadBalancingPolicy(
+    void* arg, grpc_error* error) {
+  auto* self = static_cast<LoadBalancedCall*>(arg);
+  if (self->lb_recv_trailing_metadata_ready_ != nullptr) {
+    // Set error if call did not succeed.
+    grpc_error* error_for_lb = GRPC_ERROR_NONE;
+    if (error != GRPC_ERROR_NONE) {
+      error_for_lb = error;
+    } else {
+      const auto& fields = self->recv_trailing_metadata_->idx.named;
+      GPR_ASSERT(fields.grpc_status != nullptr);
+      grpc_status_code status =
+          grpc_get_status_code_from_metadata(fields.grpc_status->md);
+      std::string msg;
+      if (status != GRPC_STATUS_OK) {
+        error_for_lb = grpc_error_set_int(
+            GRPC_ERROR_CREATE_FROM_STATIC_STRING("call failed"),
+            GRPC_ERROR_INT_GRPC_STATUS, status);
+        if (fields.grpc_message != nullptr) {
+          error_for_lb = grpc_error_set_str(
+              error_for_lb, GRPC_ERROR_STR_GRPC_MESSAGE,
+              grpc_slice_ref_internal(GRPC_MDVALUE(fields.grpc_message->md)));
+        }
+      }
+    }
+    // Invoke callback to LB policy.
+    Metadata trailing_metadata(self, self->recv_trailing_metadata_);
+    LbCallState lb_call_state(self);
+    self->lb_recv_trailing_metadata_ready_(error_for_lb, &trailing_metadata,
+                                           &lb_call_state);
```

Comment:
So this is potentially an internal 'API change' as to what this particular callback does, as the previous version had a pointer to an object to a member variable that persisted beyond the call whereas the new code passes in a pointer to an object that only lives just past the call. Aree the usage requirements of this callback documented, and are there any use cases that might have made a different assumption regarding their use?